### PR TITLE
oasp4j-jpa compatiblity for DAO with paging support

### PR DESCRIPTION
**Pull request by vapadwal**
Please find the PR for issue https://github.com/oasp/oasp4j/issues/669

### CODE REVIEW

**Comment by vapadwal** [comment]
Please also see mtsj PR https://github.com/oasp/my-thai-star/pull/128

File: modules/jpa/pom.xml [hunk 1]
```diff
@@ -19,6 +19,18 @@
...
+     <groupId>io.oasp.java.modules</groupId>
+  <artifactId>oasp4j-jpa-envers</artifactId>
+      </dependency>
+   <dependency>
+      <groupId>org.hibernate</groupId>
+      <artifactId>hibernate-envers</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.springframework</groupId>
+    <artifactId>spring-context</artifactId>
```
Comment (hohwille) [inline modules/jpa/pom.xml]: nice to have would be a formatting ([Ctrl][Shift][f]) as the indentation is a little mixed.

File: modules/jpa/pom.xml [hunk 2]
```diff
@@ -19,6 +19,18 @@
       <groupId>io.oasp.java.modules</groupId>
       <artifactId>oasp4j-jpa-dao</artifactId>
     </dependency>
+    <dependency>
+     <groupId>io.oasp.java.modules</groupId>
+  <artifactId>oasp4j-jpa-envers</artifactId>
```
Comment (hohwille) [inline modules/jpa/pom.xml]: why do we need `oasp4j-jpa-envers` here? Wasn't the dependency other-way round before?

File: modules/jpa/pom.xml [hunk 3]
```diff
@@ -19,6 +19,18 @@
...
+      </dependency>
+   <dependency>
+      <groupId>org.hibernate</groupId>
+      <artifactId>hibernate-envers</artifactId>
```
Comment (hohwille) [inline modules/jpa/pom.xml]: Aint this a transitive dependency of `oasp4j-jpa-envers`?

File: modules/jpa/src/main/java/io/oasp/module/jpa/common/api/ApplicationDao.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+import io.oasp.module.jpa.dataaccess.api.Dao;
+import io.oasp.module.jpa.dataaccess.api.GenericRevisionedDao;
+import io.oasp.module.jpa.dataaccess.api.MutablePersistenceEntity;
+
+/**
+ * Interface for all {@link GenericRevisionedDao DAOs} (Data Access Object) of this application.
+ *
+ * @param <ENTITY> is the type of the managed entity.
+ */
+public interface ApplicationDao<ENTITY extends MutablePersistenceEntity<Long>> extends Dao<ENTITY> {
```
Comment (hohwille) [inline modules/jpa/src/main/java/io/oasp/module/jpa/common/api/ApplicationDao.java]: ApplicationDao is usually the one that goes to the application in order to let the project add its own bells and whistles there. Having this in our module would IMHO cause namespace issues with CobiGen that generates another `ApplicationDao` inside the app and then cannot resolve the imports automatically anymore as the name is ambiguous.

File: modules/jpa/src/main/java/io/oasp/module/jpa/common/base/AbstractDao.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ * Abstract base implementation of {@link Dao} interface.
+ *
+ * @param <E> is the generic type of the {@link PersistenceEntity}. Should be derived from
+ *        {@link io.oasp.module.jpa.dataaccess.api.AbstractPersistenceEntity}.
+ *
+ * @deprecated It is only provided to support old functionality of DAO, instead use
+ *             io.oasp.module.jpa.dataaccess.base.AbstractDao instead
+ */
+@Deprecated
+public abstract class AbstractDao<E extends PersistenceEntity<Long>> extends AbstractGenericDao<Long, E>
```
Comment (hohwille) [inline modules/jpa/src/main/java/io/oasp/module/jpa/common/base/AbstractDao.java]: This looks good and deprecation is fine.

File: modules/jpa/src/main/java/io/oasp/module/jpa/common/base/AbstractGenericDao.java [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+ *
+ * @param <ID> is the generic type if the {@link PersistenceEntity#getId() primary key}.
+ * @param <E> is the generic type of the managed {@link PersistenceEntity}.
+ *
+ * @deprecated It is only provided to support old functionality of DAO, instead use
+ *             io.oasp.module.jpa.dataaccess.base.AbstractGenericDao instead
+ */
+// @Repository
+@Deprecated
+public abstract class AbstractGenericDao<ID, E extends PersistenceEntity<ID>> implements GenericDao<ID, E> {
```
Comment (hohwille) [inline modules/jpa/src/main/java/io/oasp/module/jpa/common/base/AbstractGenericDao.java]: This is also fine in general but I would expect this to extend another abstract class from `oasp4j-jpa-dao` containing the general DAO implementation stuff and only having the legacy support methods (paging support, etc.) here.

File: modules/jpa/src/main/java/io/oasp/module/jpa/common/base/AbstractGenericRevisionedDao.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+ *
+ * @param <ID> is the type of the {@link MutablePersistenceEntity#getId() primary key} of the managed
+ *        {@link MutablePersistenceEntity entity}.
+ * @param <ENTITY> is the {@link #getEntityClass() type} of the managed entity.
+ *
+ * @deprecated It is only provided to support old functionality of DAO, instead use
+ *             io.oasp.module.jpa.dataaccess.base.AbstractGenericRevisionedDao instead
+ */
+@Deprecated
+public abstract class AbstractGenericRevisionedDao<ID, ENTITY extends MutablePersistenceEntity<ID>>
```
Comment (hohwille) [inline modules/jpa/src/main/java/io/oasp/module/jpa/common/base/AbstractGenericRevisionedDao.java]: This class was IMHO before present in `oasp4j-jpa-envers` and I would still expect it there. This is completely envers specific.

File: modules/jpa/src/main/java/io/oasp/module/jpa/common/base/AbstractGenericDao.java [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+/**
+ * This is the abstract base-implementation of the {@link GenericDao} interface.
+ *
+ * @param <ID> is the generic type if the {@link PersistenceEntity#getId() primary key}.
+ * @param <E> is the generic type of the managed {@link PersistenceEntity}.
+ *
+ * @deprecated It is only provided to support old functionality of DAO, instead use
+ *             io.oasp.module.jpa.dataaccess.base.AbstractGenericDao instead
+ */
+// @Repository
```
Comment (hohwille) [inline modules/jpa/src/main/java/io/oasp/module/jpa/common/base/AbstractGenericDao.java]: Please remove this line.

File: modules/jpa/src/main/java/io/oasp/module/jpa/common/base/AbstractGenericRevisionedDao.java (hunk 7 above)
Comment (vapadwal) [reply modules/jpa/src/main/java/io/oasp/module/jpa/common/base/AbstractGenericRevisionedDao.java]: Basically what i found below hierarchy here in mtsj that
 BookingDaoImpl->ApplicationDaoImpl->AbstractRevisionedDao->AbstractGenericRevisionedDao->AbstractGenericDao

So I tried to replicate this one, if i tried to use oasp4j-jpa-envers than it would call AbstractGenericDao from oasp4j-jpa-dao 

So i prepared BookingDaoImpl->LegacyApplicationDaoImpl-->AbstractRevisionedDao->AbstractGenericRevisionedDao->AbstractGenericDao (all from oasp4j-jpa)

I thought of directly connecting from from LegacyApplicationDaoImpl to AbstractGenericDao  but not aware of the implication of skipping  oasp4j-jpa-envers

So basically now BookingDaoImpl extends LegacyApplicationDaoImpl and can use findPaginated method as it is and would required minimal changes in mtsj

File: modules/jpa/src/main/java/io/oasp/module/jpa/common/base/AbstractGenericDao.java (hunk 8 above)
Comment (vapadwal) [reply modules/jpa/src/main/java/io/oasp/module/jpa/common/base/AbstractGenericDao.java]: done

File: modules/jpa/src/main/java/io/oasp/module/jpa/common/base/AbstractGenericDao.java (hunk 6 above)
Comment (vapadwal) [reply modules/jpa/src/main/java/io/oasp/module/jpa/common/base/AbstractGenericDao.java]: This can be done all Paging support method put in separate class

File: modules/jpa/src/main/java/io/oasp/module/jpa/common/api/ApplicationDao.java (hunk 4 above)
Comment (vapadwal) [reply modules/jpa/src/main/java/io/oasp/module/jpa/common/api/ApplicationDao.java]: As LegacyApplicationDaoImpl  is implementing it , its name can be changed to LegacyApplicationDao

File: modules/jpa/pom.xml (hunk 3 above)
Comment (vapadwal) [reply modules/jpa/pom.xml]: done

File: modules/jpa/pom.xml (hunk 2 above)
Comment (vapadwal) [reply modules/jpa/pom.xml]: please see comments given to AbstractGenericRevisionedDao

**Comment by hohwille** [comment]
Here is my design suggestion. Indeed quite complicated. However, when removing/ignoring the deprecated legacy stuff, the official things do not get any more complex. And this way we avoid namespace clashing when multiple oasp4j JARs come together on the classpath.

oasp4j-dao-legacy-support

**Comment by vapadwal** [comment]
Regarding renaming Dao interfaces in oasp-jpa-dao and oasp-jpa-dao-envers had prefix "I" and  Dao abstract class have put suffix "Impl"

**Comment by hohwille** [comment]
> Regarding renaming Dao interfaces in oasp-jpa-dao and oasp-jpa-dao-envers had prefix "I"

https://github.com/oasp/oasp4j/wiki/coding-conventions#naming

Did you read the `IMPORTANT:` note? IMHO there is no need to duplicate the interfaces. Also the Envers-Classes are not DAO specific and should IMHO not have been moved to a dao specific module. What if someone wants to use envers with `oasp4j-jpa-spring-data` instead? Then he would also have to depend on `oasp4j-jpa-dao-envers` what does not make sense to me.

**Comment by hohwille** [comment]
As we are running out of time I will try to create another PR and give it a try. Then we can match and compare, discuss, pick and choose. It seems to complicated to express all in words without misunderstandings (and between the few time slots I do have).

**Comment by hohwille** [comment]
Community voted for #675 so I am closing this PR.
